client/src/service.rs

```diff
@@ -122,39 +122,48 @@ where
         self.machine.attempted(addr)
     }
 
-    fn connected(&mut self, addr: net::SocketAddr, local_addr: &net::SocketAddr, link: Link) {
+    fn connected(
+        &mut self,
+        addr: net::SocketAddr,
+        local_addr: &net::SocketAddr,
+        link: ConnDirection,
+    ) {
         self.inboxes.insert(addr, p2p::stream::Decoder::new(1024));
         self.machine.connected(addr, local_addr, link)
     }
 
     fn disconnected(
         &mut self,
         addr: &net::SocketAddr,
-        reason: DisconnectReason<Self::DisconnectReason>,
+        reason: DisconnectReason<Self::DisconnectDemand>,
     ) {
         self.inboxes.remove(addr);
         self.machine.disconnected(addr, reason)
     }
 }
 
 impl<T, F, P, C> Iterator for Service<T, F, P, C> {
-    type Item = Io<Vec<u8>, p2p::Event, p2p::DisconnectReason>;
+    type Item = ReactorDispatch<Vec<u8>, p2p::Event, p2p::DisconnectReason>;
 
     fn next(&mut self) -> Option<Self::Item> {
         match self.machine.next() {
-            Some(Io::Write(addr, msg)) => {
```

Comment:
I'm worried if we arrive an one point where AsyncIO will be implemented in rust, this code need to be reverted because make more sense in this way, but no strong opinion on it